File: artifacts/api-server/src/routes/reports.ts
import { Router, type IRouter } from "express";
import { eq, and } from "drizzle-orm";
import { db, reportsTable, blocksTable, usersTable } from "@workspace/db";
import { CreateReportBody, BlockUserBody, UnblockUserParams } from "@workspace/api-zod";

const router: IRouter = Router();

router.post("/reports", async (req, res): Promise<void> => {
  if (!req.session.userId) {
    res.status(401).json({ error: "Not authenticated" });
    return;
  }

  const parsed = CreateReportBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const { reportedUserId, reason, description } = parsed.data;

  await db.insert(reportsTable).values({
    reporterId: req.session.userId,
    reportedUserId,
    reason,
    description: description ?? null,
  });

  res.status(201).json({ message: "Report submitted successfully" });
});

router.get("/blocks", async (req, res): Promise<void> => {
  if (!req.session.userId) {
    res.status(401).json({ error: "Not authenticated" });
    return;
  }

  const blocks = await db.select().from(blocksTable).where(eq(blocksTable.blockerId, req.session.userId));

  const result = await Promise.all(blocks.map(async (block) => {
    const [blocked] = await db.select().from(usersTable).where(eq(usersTable.id, block.blockedUserId));
    return {
      id: block.id,
      blockedUserId: block.blockedUserId,
      blockedUserName: blocked?.name ?? "Unknown",
      createdAt: block.createdAt.toISOString(),
    };
  }));

  res.json(result);
});

router.post("/blocks", async (req, res): Promise<void> => {
  if (!req.session.userId) {
    res.status(401).json({ error: "Not authenticated" });
    return;
  }

  const parsed = BlockUserBody.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: parsed.error.message });
    return;
  }

  const { blockedUserId } = parsed.data;

  if (blockedUserId === req.session.userId) {
    res.status(400).json({ error: "Cannot block yourself" });
    return;
  }

  const existing = await db.select().from(blocksTable).where(
    and(eq(blocksTable.blockerId, req.session.userId), eq(blocksTable.blockedUserId, blockedUserId))
  );

  if (existing.length > 0) {
    res.status(201).json({ message: "Already blocked" });
    return;
  }

  await db.insert(blocksTable).values({
    blockerId: req.session.userId,
    blockedUserId,
  });

  res.status(201).json({ message: "User blocked successfully" });
});

router.delete("/blocks/:blockedUserId", async (req, res): Promise<void> => {
  if (!req.session.userId) {
    res.status(401).json({ error: "Not authenticated" });
    return;
  }

  const params = UnblockUserParams.safeParse(req.params);
  if (!params.success) {
    res.status(400).json({ error: params.error.message });
    return;
  }

  await db.delete(blocksTable).where(
    and(eq(blocksTable.blockerId, req.session.userId), eq(blocksTable.blockedUserId, params.data.blockedUserId))
  );

  res.sendStatus(204);
});

export default router;

File: Python/routers/reports.py
from fastapi import APIRouter, Request, HTTPException, Response
from pydantic import BaseModel
from typing import Optional
from db import db

router = APIRouter()

class CreateReportBody(BaseModel):
    reportedUserId: int
    reason: str
    description: Optional[str] = None

class BlockUserBody(BaseModel):
    blockedUserId: int

@router.post("/reports", status_code=201)
def create_report(body: CreateReportBody, request: Request):
    user_id = request.session.get("user_id")
    if not user_id:
        raise HTTPException(status_code=401, detail="Not authenticated")

    with db() as conn:
        cur = conn.cursor()
        cur.execute(
            "INSERT INTO reports (reporter_id, reported_user_id, reason, description) VALUES (%s, %s, %s, %s)",
            (user_id, body.reportedUserId, body.reason, body.description),
        )
    return {"message": "Report submitted successfully"}

@router.get("/blocks")
def list_blocks(request: Request):
    user_id = request.session.get("user_id")
    if not user_id:
        raise HTTPException(status_code=401, detail="Not authenticated")

    with db() as conn:
        cur = conn.cursor()
        cur.execute("SELECT * FROM blocks WHERE blocker_id = %s", (user_id,))
        blocks = cur.fetchall()
        result = []
        for block in blocks:
            cur.execute("SELECT name FROM users WHERE id = %s", (block["blocked_user_id"],))
            blocked_user = cur.fetchone()
            result.append({
                "id": block["id"],
                "blockedUserId": block["blocked_user_id"],
                "blockedUserName": blocked_user["name"] if blocked_user else "Unknown",
                "createdAt": block["created_at"].isoformat(),
            })
        return result

@router.post("/blocks", status_code=201)
def block_user(body: BlockUserBody, request: Request):
    user_id = request.session.get("user_id")
    if not user_id:
        raise HTTPException(status_code=401, detail="Not authenticated")
    if body.blockedUserId == user_id:
        raise HTTPException(status_code=400, detail="Cannot block yourself")

    with db() as conn:
        cur = conn.cursor()
        cur.execute(
            "SELECT id FROM blocks WHERE blocker_id = %s AND blocked_user_id = %s",
            (user_id, body.blockedUserId),
        )
        if cur.fetchone():
            return {"message": "Already blocked"}
        cur.execute(
            "INSERT INTO blocks (blocker_id, blocked_user_id) VALUES (%s, %s)",
            (user_id, body.blockedUserId),
        )
    return {"message": "User blocked successfully"}

@router.delete("/blocks/{blocked_user_id}", status_code=204)
def unblock_user(blocked_user_id: int, request: Request):
    user_id = request.session.get("user_id")
    if not user_id:
        raise HTTPException(status_code=401, detail="Not authenticated")

    with db() as conn:
        cur = conn.cursor()
        cur.execute(
            "DELETE FROM blocks WHERE blocker_id = %s AND blocked_user_id = %s",
            (user_id, blocked_user_id),
        )
